client/jetpack-cloud/sections/partner-portal/hooks.tsx

```diff
@@ -0,0 +1,281 @@
+			}
+
+			if ( newPage < 1 ) {
+				return;
+			}
+
+			if ( direction === 'next' && ! hasMore ) {
+				return;
+			}
+
+			setPage( newPage );
+			onNavigateCallback?.( newPage, direction );
+		},
+		[ enabled, hasMore, setPage, onNavigateCallback ]
+	);
+
+	return [ page, showPagination, onNavigate ];
+}
+
+/**
+ * Handle license issuing
+ *
+ */
+export function useLicenseIssuing(
+	selectedSite?: { ID: number; domain: string } | null,
+	product?: string | null
+): [
+	UseMutationResult< APILicense, APIError, MutationIssueLicenseVariables, unknown >,
+	boolean,
+	() => void
+] {
+	const translate = useTranslate();
+	const dispatch = useDispatch();
+	const sites = useSelector( getSites ).length;
+	const [ isSubmitting, setIsSubmitting ] = useState( false );
+	const products = useProductsQuery( {
+		select: alphabeticallySortedProductOptions,
+	} );
+
+	const paymentMethodRequired = useSelector( doesPartnerRequireAPaymentMethod );
+	const hasPaymentMethod = useSelector( hasValidPaymentMethod );
+
+	const fromDashboard = getQueryArg( window.location.href, 'source' ) === 'dashboard';
+
+	const handleRedirectToDashboard = ( licenseKey: string ) => {
+		const selectedProduct = products?.data?.find( ( p ) => p.slug === product );
+		if ( selectedSite && selectedProduct ) {
+			dispatch(
+				setPurchasedLicense( {
+					selectedSite: selectedSite?.domain,
+					selectedProduct: {
+						name: getProductTitle( selectedProduct.name ),
+						key: licenseKey,
+					},
+				} )
+			);
+		}
+		return page.redirect( '/dashboard' );
+	};
+
+	const requirePaymentMethod = useCallback( () => {
+		if ( paymentMethodRequired && ! hasPaymentMethod ) {
+			const nextStep = addQueryArgs(
+				{
+					product,
+				},
+				partnerPortalBasePath( '/payment-methods/add' )
+			);
+			page( nextStep );
+		}
+	}, [ product, hasPaymentMethod, paymentMethodRequired ] );
+
+	const assignLicense = useAssignLicenseMutation( {
+		onSuccess: ( license: any ) => {
+			setIsSubmitting( false );
+
+			if ( fromDashboard ) {
+				handleRedirectToDashboard( license.license_key );
+				return;
+			}
+
+			page.redirect(
+				addQueryArgs( { highlight: license.license_key }, partnerPortalBasePath( '/licenses' ) )
+			);
+		},
+		onError: ( error: Error ) => {
+			setIsSubmitting( false );
+			dispatch( errorNotice( error.message ) );
+		},
+	} );
+
+	const issueLicense = useIssueLicenseMutation( {
+		onSuccess: ( license ) => {
+			const licenseKey = license.license_key;
+			const selectedSiteId = selectedSite?.ID;
+
+			if ( selectedSiteId ) {
+				setIsSubmitting( true );
+				assignLicense.mutate( { licenseKey, selectedSite: selectedSiteId } );
+				return;
+			}
+
+			let nextStep = addQueryArgs(
+				{ highlight: license.license_key },
+				partnerPortalBasePath( '/licenses' )
+			);
+
+			if ( sites > 0 ) {
+				nextStep = addQueryArgs(
+					{ key: license.license_key },
+					partnerPortalBasePath( '/assign-license' )
+				);
+			}
+
+			page.redirect( nextStep );
```

Comment:
Removed the case that the user is redirected to `/payment-methods/add` as the following step, since this is not a real case anymore. Using `requirePaymentMethod()` on the component level will lead the user to `/payment-methods/add` before issuing a license.